src/ops/PromoteOps.ts

```diff
@@ -0,0 +1,1263 @@
+    const fileDiffname = 'fileDiff.config.json';
+    compareDirectories(exportDir, masterDir);
+
+    const compareObj: CompareObj = {
+      added: added,
+      changed: changed,
+      deleted: deleted,
+    };
+    if (printDiff) {
+      saveJsonToFile(compareObj, getFilePath('a1' + fileDiffname, true));
+    }
+
+    verboseMessage(realms);
+
+    for (const realm of realms) {
+      let realmAdded = new Array<string>();
+      let realmChanged = new Array<string>();
+      let realmDeleted = new Array<string>();
+
+      if (realm === 'global') {
+        realmAdded = added.filter((val) =>
+          val.substring(0, val.indexOf('/')).includes('global')
+        );
+        realmChanged = changed.filter((val) =>
+          val.substring(0, val.indexOf('/')).includes('global')
+        );
+        realmDeleted = deleted.filter((val) =>
+          val.substring(0, val.indexOf('/')).includes('global')
+        );
+      } else {
+        realmAdded = added.filter(
+          (val) =>
+            val.substring(
+              val.indexOf('/') + 1,
+              val.indexOf('/', val.indexOf('/') + 1)
+            ) === realm
+        );
+        realmChanged = changed.filter(
+          (val) =>
+            val.substring(
+              val.indexOf('/') + 1,
+              val.indexOf('/', val.indexOf('/') + 1)
+            ) == realm
+        );
+        realmDeleted = deleted.filter(
+          (val) =>
+            val.substring(
+              val.indexOf('/') + 1,
+              val.indexOf('/', val.indexOf('/') + 1)
+            ) === realm
+        );
+      }
+
+      verboseMessage(realm);
+      const compObj: CompareObj = {
+        added: realmAdded,
+        changed: realmChanged,
+        deleted: realmDeleted,
+      };
+      verboseMessage(compObj);
+      if (!whatIf) {
+        await effectDifferences(compObj, masterDir, exportDir, effectSecrets);
+      }
+    }
+
+    if (!whatIf) {
+      const globalSync = changed.find((val) => val === 'global/idm/sync.json');
+      if (globalSync) {
+        await changeFile('global/idm/sync.json', masterDir);
+      }
+      if (enviornmentChanged(compareObj) && effectSecrets) {
+        await applyUpdates(wait);
+        verboseMessage(
+          'Must wait around 10 minutes because the enviornment is updating'
+        );
+      }
+    }
+
+    if (printDiff) {
+      saveJsonToFile(logmessages, getFilePath('a2' + fileDiffname, true));
+    }
+
+    return true;
+  } catch (error) {
+    printError(error);
+  }
+  return false;
+}
+
+function enviornmentChanged(files: CompareObj): boolean {
+  // variables
+  let variable = files.changed.find((val) => val.includes('global/variable'));
+  if (variable) {
+    return true;
+  }
+  variable = files.added.find((val) => val.includes('global/variable'));
+  if (variable) {
+    return true;
+  }
+  variable = files.deleted.find((val) => val.includes('global/variable'));
+  if (variable) {
+    return true;
+  }
+  // secrets
```

Comment:
Why aren't secrets checked?

Reply:
They are hashed when exported so its difficult to compare them; Thinking about how to proceed with this